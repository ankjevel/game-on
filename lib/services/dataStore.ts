import uuid from 'uuid'
import redis from '../adapters/redis'
import parse from '../utils/parse'

import {
  Action,
  CreateGroupInput,
  CreateUserInput,
  GetResult,
  Group,
  isAction,
  isGroup,
  isUser,
  StoreTypes,
  User,
} from '../types/dataStore'
export { StoreTypes, User, Group, CreateGroupInput, CreateUserInput, GetResult }

const client = redis()

const newId = (type: StoreTypes) => `${type}:${uuid.v4()}`

export const create = async <T extends Group | Action | User>(
  type: StoreTypes,
  format: (id: string) => T
): Promise<T> => {
  const id = newId(type)

  if (await client.get(id)) {
    return await create(type, format)
  }

  const object = format(id)

  await client.set(
    id,
    typeof object === 'string' ? object : JSON.stringify(object)
  )

  return object
}

export const update = async <T extends Group | Action | User>(
  id: T['id'],
  data: T
) => {
  if (id !== data.id) {
    return
  }

  const prev = await get({ id })

  if (prev == null) {
    return
  }

  await client.set(id, {
    ...prev,
    ...data,
  })
}

export const get = async ({
  type,
  id,
}: {
  type?: StoreTypes
  id: string
}): Promise<GetResult> => {
  const query =
    id.includes(':') === false
      ? type == null
        ? null
        : `${type}:${id}`
      : type != null && id.split(':')[0] != type
      ? null
      : id

  if (query == null) {
    return null
  }

  const result = await client.get(query)

  if (result === null) {
    return null
  }

  try {
    if (type === StoreTypes.User || id.includes(StoreTypes.User)) {
      const parsed = parse<User>(result)
      return isUser(parsed) ? parsed : null
    }

    if (type === StoreTypes.Group || id.includes(StoreTypes.Group)) {
      const parsed = parse<Group>(result)
      return isGroup(parsed) ? parsed : null
    }

    if (type === StoreTypes.Action || id.includes(StoreTypes.Action)) {
      const parsed = parse<Action>(result)
      return isAction(parsed) ? parsed : null
    }
  } catch (error) {
    console.error(result, error)
    return null
  }

  return null
}
