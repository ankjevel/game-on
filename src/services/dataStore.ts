type Check = <T>(result: MaybeNull<T>) => boolean

import uuid from 'uuid'
import * as userService from './user'
import { parse } from '../utils'
import redis from '../adapters/redis'

import {
  Action,
  GetResult,
  Group,
  isAction,
  isGroup,
  isUser,
  StoreTypes,
  Types,
  User,
} from '../types/dataStore'

export { Action, GetResult, Group, StoreTypes, Types, User, Check }

const client = redis()

export const newId = (type: StoreTypes) => `${type}:${uuid.v4()}`

export const stripTag = (input: string) => input.split(':')[1] || ''

export const checkDuplicate = async <T extends Types>(input: T) => {
  if (isAction(input) || isGroup(input)) {
    return
  }

  if (isUser(input)) {
    return userService.checkDuplicate(input)
  }
}

export const create = async <T extends Types>(
  type: StoreTypes,
  format: (id: string) => T
): Promise<T> => {
  const id = newId(type)

  if (await client.get(id)) {
    return await create(type, format)
  }

  const object = format(id)

  await checkDuplicate<T>(object)

  await client.set(
    id,
    typeof object === 'string' ? object : JSON.stringify(object)
  )

  return object
}

export const all = async (prefix: StoreTypes) =>
  await client.keys(`${prefix}:*`)

export const update = async <T extends Types>(
  id: T['id'],
  data: T,
  type: StoreTypes
) => {
  if (id !== data.id) {
    return
  }

  const prev = await getWrapper({ id, type })
  if (prev == null) {
    return
  }

  await client.set(
    id,
    JSON.stringify({
      ...prev,
      ...data,
    })
  )
}

export const get = async <T extends Types>({
  type,
  id,
  check,
}: {
  type: StoreTypes
  id: string
  check: Check
}): Promise<MaybeNull<T>> => {
  const query =
    id.includes(':') === false
      ? type == null
        ? null
        : `${type}:${id}`
      : type != null && id.split(':')[0] != type
      ? null
      : id

  if (query == null || !query.startsWith(`${type}:`)) {
    return null
  }

  const result = await client.get(query)
  if (result == null) {
    return null
  }

  try {
    const parsed = parse<T>(result)
    return check(parsed) ? parsed : null
  } catch (error) {
    console.error(result, error)
    return null
  }
}

export const getWrapper = async <T extends Types>({
  type,
  id,
}: {
  type: StoreTypes
  id: string
}): Promise<MaybeNull<T>> => {
  let check: Check

  switch (type) {
    case StoreTypes.Action:
      check = isAction
      break
    case StoreTypes.Group:
      check = isGroup
      break
    case StoreTypes.User:
      check = isUser
      break
    default:
      throw new Error('missing type')
  }

  return get<T>({ id, type, check })
}
