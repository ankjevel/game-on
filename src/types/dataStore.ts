import { hasProp, looksLikeEmail, isNumber, nullOrEmpty } from '../utils'
import {
  Action,
  IActionRunning,
  Group,
  User,
  UserWithOutPassword,
  GetResult,
  KeyValue,
} from 'dataStore'

export type Types = Action | Group | User

export enum StoreTypes {
  Action = 'action',
  Group = 'group',
  User = 'user',
  ActionRunning = 'action:running',
}

export type ActionRunning = IActionRunning<NewAction, UserSummary>

export enum NewActionEnum {
  None = 'none',
  Bet = 'bet',
  Check = 'check',
  Call = 'call',
  Raise = 'raise',
  AllIn = 'allIn',
  Fold = 'fold',

  Draw = 'draw',
  Winner = 'winner',
  Back = 'back',
  Bank = 'bank',
  Join = 'join',
  Leave = 'leave',
  SittingOut = 'sittingOut',
}

export type NewAction = {
  type: NewActionEnum
  value?: number
  winners?: User['id'][]
}

export type UserSummary = {
  bet: number
  status: NewActionEnum
}

export const isUserSummary = (any: any): any is UserSummary =>
  any != null &&
  hasProp<any>(any, 'bet') &&
  isNumber(any.bet) &&
  hasProp<any>(any, 'status') &&
  isNewActionType((any as any).status)

export const isNewActionType = (any: any): any is NewActionEnum =>
  Object.values(NewActionEnum).includes(any)

export const isNewAction = (
  any: NewAction | unknown,
  strict = false
): any is NewAction => {
  if (
    !hasProp<NewAction>(any, 'type') ||
    !isNewActionType((any as any).type) ||
    (hasProp<NewAction>(any, 'value') && any.value != null
      ? !isNumber((any as any).value)
      : false)
  ) {
    return false
  }

  if (!strict) {
    return true
  }

  switch (any.type) {
    case NewActionEnum.AllIn:
    case NewActionEnum.Back:
    case NewActionEnum.Bet:
    case NewActionEnum.Call:
    case NewActionEnum.Check:
    case NewActionEnum.Fold:
    case NewActionEnum.None:
    case NewActionEnum.SittingOut:
      return isNumber(any.value) === false
    case NewActionEnum.Draw: // draw COULD include users that went all in.
    case NewActionEnum.Winner:
      return Array.isArray(any.winners)
        ? any.winners.every(winner => checkId(winner, StoreTypes.User))
        : true
    default:
      return isNumber(any.value)
  }
}

export { Action, Group, User, GetResult }

export const checkId = (input: string, type: StoreTypes) =>
  input.split(':')[0] === type.split(':')[0]

export const isUser = (any: User | unknown): any is User =>
  any != null &&
  (hasProp<User>(any, 'id') &&
    !nullOrEmpty(any.id) &&
    checkId(any.id, StoreTypes.User)) &&
  (hasProp(any, 'email') &&
    !nullOrEmpty(any.email) &&
    looksLikeEmail(any.email)) &&
  (hasProp(any, 'password') && !nullOrEmpty(any.password))

export const isUserWithOutPassword = (
  any: UserWithOutPassword | unknown
): any is UserWithOutPassword =>
  any != null &&
  (hasProp<UserWithOutPassword>(any, 'id') &&
    !nullOrEmpty(any.id) &&
    checkId(any.id, StoreTypes.User)) &&
  (hasProp(any, 'email') &&
    !nullOrEmpty(any.email) &&
    looksLikeEmail(any.email))

export const isGroup = (any: Group | unknown): any is Group =>
  any != null &&
  (hasProp<Group>(any, 'id') &&
    typeof any.id === 'string' &&
    !nullOrEmpty(any.id) &&
    checkId(any.id, StoreTypes.Group)) &&
  (hasProp(any, 'name') &&
    typeof any.name === 'string' &&
    !nullOrEmpty(any.name)) &&
  (hasProp(any, 'owner') &&
    typeof any.owner === 'string' &&
    !nullOrEmpty(any.owner)) &&
  (hasProp(any, 'startSum') && isNumber(any.startSum)) &&
  (hasProp(any, 'users')
    ? Array.isArray(any.users) &&
      (any.users as Group['users']).every(
        ({ id, sum }) => checkId(id, StoreTypes.User) && isNumber(sum)
      )
    : true) &&
  (hasProp(any, 'action')
    ? typeof any.action === 'string' && checkId(any.action, StoreTypes.Action)
    : true)

export const isAction = (any: Action | unknown): any is Action =>
  any != null &&
  hasProp<Action>(any, 'id') &&
  typeof any.id === 'string' &&
  checkId(any.id, StoreTypes.Action)

export const isTurn = (
  any: KeyValue<NewActionEnum> | unknown
): any is KeyValue<NewActionEnum> =>
  any != null &&
  Object.entries(any as any).every(
    ([key, value]) => checkId(key, StoreTypes.User) && isNewAction(value)
  )

export const isActionRunning = (
  any: ActionRunning | unknown
): any is ActionRunning =>
  isAction(any) &&
  (hasProp<ActionRunning>(any, 'round') && isNumber(any.round)) &&
  (hasProp(any, 'groupID') && checkId(any.groupID, StoreTypes.Group)) &&
  (hasProp(any, 'queued') && isTurn(any.queued)) &&
  (hasProp(any, 'turn') &&
    Object.entries(any.turn).every(
      ([key, value]) => checkId(key, StoreTypes.User) && isUserSummary(value)
    )) &&
  (hasProp(any, 'button') && checkId(any.button, StoreTypes.User)) &&
  (hasProp(any, 'big') && checkId(any.big, StoreTypes.User)) &&
  (hasProp(any, 'pot') && isNumber(any.round)) &&
  (Array.isArray(any.sittingOut)
    ? any.sittingOut.every(user => checkId(user, StoreTypes.User))
    : true)
