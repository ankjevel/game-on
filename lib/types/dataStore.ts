import hasProp from '../utils/hasProp'
import { Action, Group, User, CreateInput, GetResult } from 'dataStore'
export { Action, Group, User, CreateInput, GetResult }

export enum StoreTypes {
  Action = 'action',
  Group = 'group',
  User = 'user',
}

export const isUser = (any: any): any is User => {
  return any != null && hasProp(any, 'id')
}

export const isGroup = (any: any): any is Group => {
  return (
    any != null &&
    hasProp(any, 'id') &&
    hasProp(any, 'users') &&
    typeof any.id === 'string' &&
    Array.isArray(any.users) &&
    any.users.every(value => isUser(value)) &&
    ((hasProp(any, 'action') && isAction(any.action)) || true)
  )
}

export const isAction = (any: any): any is Action => {
  return any != null && hasProp(any, 'id') && typeof any.id === 'string'
}
