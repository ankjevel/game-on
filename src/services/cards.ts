import {
  Card,
  Cards,
  Element,
  Hand,
  HandOrder,
  SameObject,
  Sorted,
  SortHigh,
  Suit,
  Tied,
} from 'cards'
import { MutableDeck, Deck, ActionRunning, HandParsed } from 'dataStore'

import { n, clone } from '../utils'

export enum HandEnum {
  RoyalFlush = 0,
  StraightFlush = 1,
  FourOfAKind = 2,
  FullHouse = 3,
  Flush = 4,
  Straight = 5,
  ThreeOfAKind = 6,
  TwoPair = 7,
  Pair = 8,
  HighCard = 9,
}

/**
 * Card | Suit | value
 */
export enum Enum {
  Card = 0x1f000,
  Spades = 0xa0,
  Hearts = 0xb0,
  Diamonds = 0xc0,
  Clubs = 0xd0,
}

const defaultSameObject = () =>
  clone(n(14).reduce((x, _, i) => {
    if (i === 11) return x
    x[i + 1] = 0
    return x
  }, {}) as SameObject)

const defaultFlushObject = () =>
  clone({
    spades: false,
    hearts: false,
    diamonds: false,
    clubs: false,
  })

const sort = (a: number, b: number) => (a > b ? 1 : a < b ? -1 : 0)
const sortAceLast = (a: number, b: number) =>
  a === 1 ? 1 : b === 1 ? -1 : a > b ? 1 : a < b ? -1 : 0

const getHEX = (suit: Suit): number => {
  switch (suit) {
    case 'clubs':
      return Enum.Clubs
    case 'diamonds':
      return Enum.Diamonds
    case 'hearts':
      return Enum.Hearts
    case 'spades':
      return Enum.Spades
    default:
      throw new Error('missing suit')
  }
}

const genCards = (hex: number) =>
  n(14)
    .map((_, i) => {
      if (i === 11) {
        // 11 is Knight
        return null
      }
      return (hex | (i + 1)).toString(16)
    })
    .filter(x => x) as Tuple<string, 13>

const genSuit = (suit: Suit) => {
  if (
    suit !== 'clubs' &&
    suit !== 'diamonds' &&
    suit !== 'hearts' &&
    suit !== 'spades'
  ) {
    throw new Error('missing suit')
  }

  return genCards(getHEX(suit))
}

const DECK = (['clubs', 'diamonds', 'hearts', 'spades'] as Suit[]).flatMap(
  suit => genSuit(suit)
) as Deck

export const newDeck = () => DECK.slice(0) as Deck

/**
 *
 * @param deck  Mutable parameter
 * @param take  Value between 1 and Infinity
 */
export const takeCards = <T>(deck: MutableDeck, take: number): T => {
  const result: string[] = []

  for (const _ of n(take)) {
    let cards = deck.filter(x => typeof x === 'string')

    if (cards.length < 1) {
      deck.splice(0, 0, ...DECK.slice(0))
      cards = deck
    }

    const index = Math.floor(Math.random() * (cards.length - 1))
    const card = cards[index]
    const indexOfCard = deck.indexOf(card)

    if (!card || indexOfCard === -1) {
      throw new Error('missing card!')
    }

    deck.splice(indexOfCard, 1, null)

    result.push(card)
  }

  return (result as unknown) as T
}

export const sortCards = (cards: Card[]) =>
  Object.entries(
    cards.reduce(
      (suits: Sorted, card) => {
        if ((card ^ Enum.Hearts) <= 14) {
          suits.hearts.push(card ^ Enum.Hearts)
        }
        if ((card ^ Enum.Diamonds) <= 14) {
          suits.diamonds.push(card ^ Enum.Diamonds)
        }
        if ((card ^ Enum.Clubs) <= 14) {
          suits.clubs.push(card ^ Enum.Clubs)
        }
        if ((card ^ Enum.Spades) <= 14) {
          suits.spades.push(card ^ Enum.Spades)
        }
        return suits
      },
      {
        spades: [],
        hearts: [],
        diamonds: [],
        clubs: [],
      }
    )
  ).reduce((suits, [suit, cards]) => {
    suits[suit] = cards.sort(sort)
    return suits
  }, {}) as Sorted

export const parseHand = (communityCards: Cards, hand: Cards) => {
  const cards = sortCards(
    [...communityCards, ...hand].map(card => parseInt(card, 16) as Card)
  )
  const flush = Object.entries(cards).reduce((flush, [suit, cards]) => {
    flush[suit] = cards.length >= 5
    return flush
  }, defaultFlushObject())
  const same = Object.values(cards).reduce((list, cards) => {
    cards.forEach(card => ++list[card])
    return list
  }, defaultSameObject())
  const sameEntries = Object.entries(same)
  const pairs = sameEntries
    .filter(([, amount]) => amount >= 2)
    .map(([card]) => card)
  const threeOfAKinds = sameEntries
    .filter(([, amount]) => amount >= 3)
    .map(([card]) => card)
  const fourOfAKinds = sameEntries
    .filter(([, amount]) => amount >= 4)
    .map(([card]) => card)
  const straightFlushes = Object.entries(flush)
    .filter(([suit, isFlush]) => isFlush && hasStraight(cards[suit])[0])
    .map(([suit]) => suit)

  return {
    cards,
    flush,
    same,
    pairs,
    threeOfAKinds,
    fourOfAKinds,
    straightFlushes,
  } as HandParsed['parsed']
}

const straightOrders = [
  {
    2: 1,
    3: 2,
    4: 3,
    5: 4,
    6: 5,
    7: 6,
    8: 7,
    9: 8,
    10: 9,
    11: 10,
    13: 11,
    14: 13,
  },
  {
    3: 2,
    4: 3,
    5: 4,
    6: 5,
    7: 6,
    8: 7,
    9: 8,
    10: 9,
    11: 10,
    13: 11,
    14: 13,
    1: 14,
  },
]

const sliceAndCheck = (cards: number[], index: number, orderIndex: number) => {
  const slice: number[] = cards.slice(index, cards.length).slice(0, 5)

  const hasStraight =
    slice.length >= 5 &&
    slice.every(
      (card, i, cards) =>
        i === 0 || straightOrders[orderIndex][card] === cards[i - 1]
    )

  return [hasStraight, slice]
}

export const hasStraight = (cards: number[]) => {
  let straight: number[] = []

  const hasStraight =
    cards.sort(sort).some((_, i, cards) => {
      const [hasStraight, result] = sliceAndCheck(cards, i, 0)
      if (hasStraight) {
        straight = result as any
      }
      return hasStraight
    }) ||
    cards.sort(sortAceLast).some((_, i, cards) => {
      const [hasStraight, result] = sliceAndCheck(cards, i, 1)
      if (hasStraight) {
        straight = result as any
      }
      return hasStraight
    })

  return [hasStraight, straight]
}

export const sortHands = (turn: ActionRunning['turn']) => {
  const tied: Tied[] = []
  const reduced = Object.entries(turn)
    .filter(([, summary]) => summary.status !== 'fold')
    .reduce(
      (order, [id, { handParsed }]) => {
        order[
          handParsed == null ||
          Array.isArray(handParsed.onHand) === false ||
          !handParsed.onHand.length
            ? HandEnum.HighCard
            : handParsed.onHand[0]
        ].push([id, handParsed] as any)
        return order
      },
      n(HandEnum.HighCard + 1).map(() => [] as any) as Element[]
    )
  const sorted = reduced.map((order, index) => {
    if (order.length <= 1) {
      return order
    }

    return order.sort((a: any, b: any) => {
      const sortOrder = getHandOrder(index)(a, b)

      if (sortOrder === 0) {
        const _a = order.splice(order.indexOf(a), 1).pop() as any
        const _b = order.splice(order.indexOf(b), 1).pop() as any

        if (_a == null) {
          return -1
        }

        if (_b == null) {
          return 1
        }

        const insertInto = order.length
          ? getHandOrder(index)(_a, order[0] as any)
          : index

        tied.push({
          index,
          insertInto: index + insertInto,
          data: [_a, _b],
        } as any)
      }

      return sortOrder
    })
  })

  while (tied.length > 0) {
    if (tied.length === 0) {
      break
    }

    let tie: MaybeUndefined<Tied>
    while ((tie = tied.pop())) {
      if (tie == null) {
        break
      }

      const { index, insertInto, data } = tie

      for (const element of data) {
        sorted[index].splice(sorted[index].indexOf(element as any), 1)
      }

      const next = index + 1
      const last = next === HandEnum.HighCard + 1

      if (last) {
        sorted.splice(insertInto, 0, [data] as any)
        break
      }

      const order = getHandOrder(next)(data[0], data[1])

      if (order === 0) {
        tied.push({ index: next, insertInto, data })
      } else {
        const result =
          order === -1 ? [[data[0]], [data[1]]] : [[data[1]], [data[0]]]
        sorted.splice(insertInto, 0, result as any)
        break
      }
    }
  }

  return sorted
    .filter(x => x.length)
    .reduce(
      (array, element: any) =>
        array.concat(
          element.map((inner: Element | Element[]) => {
            if (Array.isArray(inner[0])) {
              return (inner as Element[]).map(e => e[0])
            }
            return [inner[0]]
          })
        ),
      [] as string[][]
    )
}

const sortHigh: SortHigh = (a, b) => {
  const aParsed = typeof a === 'number' ? a : parseInt(a || '-1')
  const bParsed = typeof b === 'number' ? b : parseInt(b || '-1')
  const aValue = aParsed === 1 ? 15 : aParsed
  const bValue = bParsed === 1 ? 15 : bParsed
  return aValue > bValue ? -1 : bValue > aValue ? 1 : 0
}

const each = (a: any[], b: any[], check: SortHigh = sortHigh) => {
  let high: 0 | 1 | -1 = 0
  for (const i of n(Math.max(a.length, b.length))) {
    const res = check(a[i], b[i])
    if (res === 0) continue
    high = res
    break
  }
  return high
}

const getSuit = (parsed: HandParsed['parsed']) =>
  (parsed.cards[
    Object.entries(parsed.flush)
      .filter(x => x[1])
      .map(x => x[0]) as any
  ] || []) as number[]

export const getHandOrder = (hand: HandEnum): HandOrder => {
  switch (hand) {
    case HandEnum.FourOfAKind: {
      return ([, { parsed: a }], [, { parsed: b }]) =>
        each(a.fourOfAKinds, b.fourOfAKinds)
    }
    case HandEnum.FullHouse:
      return ([, { parsed: a }], [, { parsed: b }]) =>
        each(a.threeOfAKinds, b.threeOfAKinds) || each(a.pairs, b.pairs)
    case HandEnum.RoyalFlush:
    case HandEnum.StraightFlush:
    case HandEnum.Flush:
      return ([, { parsed: a }], [, { parsed: b }]) =>
        each(getSuit(a), getSuit(b))
    case HandEnum.Straight:
      return ([, { parsed: a }], [, { parsed: b }]) =>
        sortHigh(a.straightHigh, b.straightHigh)
    case HandEnum.ThreeOfAKind:
      return ([, { parsed: a }], [, { parsed: b }]) =>
        each(a.threeOfAKinds, b.threeOfAKinds)
    case HandEnum.TwoPair:
    case HandEnum.Pair:
      return ([, { parsed: a }], [, { parsed: b }]) => each(a.pairs, b.pairs)
    case HandEnum.HighCard:
      return ([, { highCards: a }], [, { highCards: b }]) => each(a, b)
    default:
      throw new Error(`Hand not defined (${hand})`)
  }
}

export const checkHand = (communityCards: Cards, hand: Cards) => {
  const parsed = parseHand(communityCards, hand)
  const cards = [...new Set(Object.values(parsed.cards).flatMap(x => x))].sort(
    sort
  )
  const highCards = cards
    .slice(0)
    .sort(sortAceLast)
    .reverse() as Card[]
  const pair = parsed.pairs.length >= 1
  const twoPair = parsed.pairs.length >= 2
  const threeOfAKind = parsed.threeOfAKinds.length >= 1
  const [hadStraight, straight] = hasStraight(cards.slice(0))
  const flush = Object.values(parsed.flush).some(flush => flush)
  const fullHouse = parsed.pairs.length >= 2 && parsed.threeOfAKinds.length >= 1
  const fourOfAKind = parsed.fourOfAKinds.length >= 1
  const straightFlush = parsed.straightFlushes.length >= 1
  const royalFlush = parsed.straightFlushes
    .map(
      suit =>
        parsed.cards[suit].filter(i => [10, 11, 13, 14, 1].includes(i))
          .length === 5
    )
    .some(bool => bool)

  parsed.straightHigh = ((straight as number[]) || []).pop() as any

  return {
    parsed,
    highCards,
    onHand: [
      [HandEnum.RoyalFlush, royalFlush],
      [HandEnum.StraightFlush, straightFlush],
      [HandEnum.FourOfAKind, fourOfAKind],
      [HandEnum.FullHouse, fullHouse],
      [HandEnum.Flush, flush],
      [HandEnum.Straight, hadStraight],
      [HandEnum.ThreeOfAKind, threeOfAKind],
      [HandEnum.TwoPair, twoPair],
      [HandEnum.Pair, pair],
    ]
      .filter(([, hasOnHand]) => hasOnHand)
      .map(([hand]) => hand) as Hand[],
  }
}

export default newDeck
