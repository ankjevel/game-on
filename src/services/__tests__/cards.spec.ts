import { Deck } from 'dataStore'
import * as cards from '../cards'

describe('#newDeck', () => {
  test('generates a new deck with expected hex-values', () => {
    const deck = cards.newDeck()

    expect(deck).toHaveLength(52)
    expect(deck).toMatchSnapshot()
    expect(deck[0]).toMatchInlineSnapshot(`"1F0A1"`)
    expect(deck[deck.length - 1]).toMatchInlineSnapshot(`"1F0D13"`)
  })
})

describe('#takeCards', () => {
  test('it returns the expected amount of cards from the deck', () => {
    const deck = cards.newDeck() as Deck
    const take = cards.takeCards(deck, 2)

    expect(take).toHaveLength(2)
    expect(deck.filter(x => x == null)).toHaveLength(2)
  })
})
