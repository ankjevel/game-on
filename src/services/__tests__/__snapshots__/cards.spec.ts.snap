// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#newDeck generates a new deck with expected hex-values 1`] = `
Array [
  "1F0A1",
  "1F0A2",
  "1F0A3",
  "1F0A4",
  "1F0A5",
  "1F0A6",
  "1F0A7",
  "1F0A8",
  "1F0A9",
  "1F0A10",
  "1F0A11",
  "1F0A12",
  "1F0A13",
  "1F0B1",
  "1F0B2",
  "1F0B3",
  "1F0B4",
  "1F0B5",
  "1F0B6",
  "1F0B7",
  "1F0B8",
  "1F0B9",
  "1F0B10",
  "1F0B11",
  "1F0B12",
  "1F0B13",
  "1F0C1",
  "1F0C2",
  "1F0C3",
  "1F0C4",
  "1F0C5",
  "1F0C6",
  "1F0C7",
  "1F0C8",
  "1F0C9",
  "1F0C10",
  "1F0C11",
  "1F0C12",
  "1F0C13",
  "1F0D1",
  "1F0D2",
  "1F0D3",
  "1F0D4",
  "1F0D5",
  "1F0D6",
  "1F0D7",
  "1F0D8",
  "1F0D9",
  "1F0D10",
  "1F0D11",
  "1F0D12",
  "1F0D13",
]
`;
